docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+
+.. _reading-items:
+
+Reading ``ItemsPerThread``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The change to reading happens inside `read_global_safe`:
+
+.. code-block:: C++
+
+    auto read_global_safe = [&](const int32_t i) -> hip::static_array<T, ItemsPerThread>
+    {
+        return [&]<int32_t... I>(std::integer_sequence<int32_t, I...>)
+        {
+            if(i + ItemsPerThread < front_size)
+                return hip::static_array<T, ItemsPerThread>{
+                    front[i + I]...
+                };
+            else
+                return hip::static_array<T, ItemsPerThread>{
+                    (i + I < front_size ? front[i + I] : zero_elem)...
+                };
+        }(std::make_integer_sequence<int32_t, ItemsPerThread>());
+    };
+
+Note that each array element is being loaded consecutively without the flexibility of a configurable ``ItemsPerThread`` property. This is morally equivalent to:
+
+.. code-block:: C++
+
+    T arr[4] = {
+        front[gid + 0],
+        front[gid + 1],
+        front[gid + 2],
+        front[gid + 3]
+    }
+
+This is exactly what's happening in the ``front[i + I]...`` fold-expression. However, this can only be issued if the entire read is operating on real input without padding using ``zero_elem``. If some reads over-index the input, the read turns into:
+
+.. code-block:: C++
+
+    T arr[4] = {
+        i + 0 < front_size ? front[i + 0] : zero_elem,
+        i + 1 < front_size ? front[i + 1] : zero_elem,
+        i + 2 < front_size ? front[i + 2] : zero_elem,
+        i + 3 < front_size ? front[i + 3] : zero_elem
+    }
+
+This is performed to make it easier for the compiler to recognize vector loads from global. As the performance at large is dominated by how you move the data, it's only natural to utilize dedicated instructions to move more data with less binary. This is evident by the huge performance improvement when loading two values per thread, See `the compiler explorer <https://godbolt.org/z/b36Eea69q>`_ to learn how loading for AMD (both RDNA and CDNA) compiles to ``global_load_dwordx4``, where ``x4`` denotes the 4-vector variant of the instruction.
+
+.. note::
+
+    Note that ``read_global_safe`` that used to take an ``uint32_t`` as the index type, takes a signed integer now. When indexing an array with unsigned integers, the compiler has to handle integer overflows, as they're defined by the C/C++ standards. It might happen, that some part of the vector load indices overflow, thus resulting in a non-contiguous read. If you change the previously linked code to use an unsigned integer as the thread id, the compiler won't emit a vector load. Signed integer overflow is an undefined behavior, and hence unknown to the optimizer. To convey the absence of overflow to the compiler with unsigned indices, add ``__builtin_assume(gid + 4 > gid)``, or the more portable ``[[assume]](gid + 4 > gid)``, once ``amdclang++`` supports it.
```

Comment:
```suggestion
    Note that ``read_global_safe``, which used to take an ``uint32_t`` as the index type, now takes a signed integer. When indexing an array with unsigned integers, the compiler has to handle integer overflows, as the C/C++ standards defined them. It might happen that some part of the vector load indices overflow, thus resulting in a non-contiguous read. If you change the previously linked code to use an unsigned integer as the thread id, the compiler won't emit a vector load. Signed integer overflow is an undefined behavior, and hence, unknown to the optimizer. To convey the absence of overflow to the compiler with unsigned indices, add ``__builtin_assume(gid + 4 > gid)``, or the more portable ``[[assume]](gid + 4 > gid)``, once ``amdclang++`` supports it.
```